### ISSUE TITLE
Performance: CPU training epoch exceeds 60 s for quantized dense-layer plus kmeans predictor with batch normalization on temporal splits

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#622]
Public `recipe.train().run()` stalls past 60 s on `backend=cpu` with the composition `.layer(8).tanh().norm(batch).qi(2).k` then `.kmeans(2).silu().norm(batch).qi(5).0` and `.loss(ce)` over `recipe.data("data/temporal/chronological_splits").target("target").split(0.8)` (~9600 training rows, 1 feature). The arithmetic budget is minimal: one dense `layer(8)` with tanh and `.norm(batch)` plus a 2-cluster `.kmeans()` predictor with `.norm(batch)` produces well under 1 MFLOP per epoch and memory traffic under 200 KB. A defensible CPU total (compilation + one epoch) is under 5 s, so 60 s is materially slower. The earliest failing operation is `recipe.train().run()` before any epoch completes; the inspected implementation path is the CPU native backend (`compile_model` / `NativeTape::epoch`) for this quantized batch-normalized dense-plus-predictor graph. No existing open performance issue covers CPU stalls through this exact layer-plus-estimator, batch-normalization, and integer-quantization path (`.qi(2).k` / `.qi(5).0`) on the temporal-split dataset; closest issues (605 nv0 small workload, 612 CPU `lgbm` split scan, 607 nv0 temporal `kmeans`/residual, 617/610 CPU `rnn`/`lstm` scan) reach different earliest implementation paths.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `2bb59a9ad18ca19d`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "Performance: CPU training epoch exceeds 60 s for quantized dense-layer plus kmeans predictor with batch normalization on temporal splits",
  "body": "Public `recipe.train().run()` stalls past 60 s on `backend=cpu` with the composition `.layer(8).tanh().norm(batch).qi(2).k` then `.kmeans(2).silu().norm(batch).qi(5).0` and `.loss(ce)` over `recipe.data(\"data/temporal/chronological_splits\").target(\"target\").split(0.8)` (~9600 training rows, 1 feature). The arithmetic budget is minimal: one dense `layer(8)` with tanh and `.norm(batch)` plus a 2-cluster `.kmeans()` predictor with `.norm(batch)` produces well under 1 MFLOP per epoch and memory traffic under 200 KB. A defensible CPU total (compilation + one epoch) is under 5 s, so 60 s is materially slower. The earliest failing operation is `recipe.train().run()` before any epoch completes; the inspected implementation path is the CPU native backend (`compile_model` / `NativeTape::epoch`) for this quantized batch-normalized dense-plus-predictor graph. No existing open performance issue covers CPU stalls through this exact layer-plus-estimator, batch-normalization, and integer-quantization path (`.qi(2).k` / `.qi(5).0`) on the temporal-split dataset; closest issues (605 nv0 small workload, 612 CPU `lgbm` split scan, 607 nv0 temporal `kmeans`/residual, 617/610 CPU `rnn`/`lstm` scan) reach different earliest implementation paths.",
  "rationale": "The reproduction reaches the earliest public failure (`recipe.train().run()` crossing the slow-runtime threshold) through the CPU native artifact and epoch execution path (`compile_model` -> `NativeTape::new` / `tape.epoch`), not through NVIDIA `ptxas` (605, 607, 618) or tree-estimator split scanning (`fit_lightgbm` in 612). The dataset (`data/temporal/chronological_splits`, 12001 rows, split 0.8) and model (`layer(8)` dense block plus `kmeans(2)` predictor, both with `.norm(batch)` and quantized integer kernels `.qi(2).k` / `.qi(5).0`) justify sub-5-second throughput on CPU; the measured 60 s gap is dominated by backend overhead for the quantized batch-normalized graph rather than arithmetic, memory traffic, or loader time. Because no existing issue represents this exact CPU dense-plus-estimator batch-normalization stall, the packet introduces a new earliest public defect rather than expanding 605, 607, 610, 611, 612, or 617."
}
```

## Deterministic failure packet

kind=performance
id=2bb59a9ad18ca19d
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:22780 next:22781 composition:48309661608500
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-6667-22780/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.layer(8).tanh().norm(batch).qi(2).k
		.kmeans(2).silu().norm(batch).qi(5).0
		.loss(ce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(48309661608500)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.bf(16)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Measured 60 s (`backend=cpu`) for one epoch of `recipe.train().run()` with model `.conv(8, 1).exp().qi(4).k .attn(1).relu().norm(batch) .loss(huber)` over `recipe.data("data/temporal/chronological_splits").target("target")`. The dataset is ~12000 rows with 1 feature (`input`); arithmetic work is minimal (8-filter kernel-1 conv, 1-head attention, batch norm) and memory traffic is under a few hundred KB, so a defensible CPU epoch should be under 5 s. The 60 s gap reaches the same earliest public failure (`recipe.train().run()` crossing the slow-runtime threshold) through the same CPU native-backend path (`compile_model` -> `NativeTape::epoch`) for a quantized batch-normalized graph covered by issue 622. The `.conv(8, 1)` and `.attn(1)` blocks, `.qi(4).k`, `.norm(batch)`, `.f(6, 9)`, and `.loss(huber)` are configuration differences that execute after the failure point; they expand the existing quantized batch-normalization stall rather than establishing a different earliest cause.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `a501e03f4f858e53`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 622,
  "title": "Performance: CPU training epoch exceeds 60 s for quantized dense-layer plus kmeans predictor with batch normalization on temporal splits",
  "body": "Measured 60 s (`backend=cpu`) for one epoch of `recipe.train().run()` with model `.conv(8, 1).exp().qi(4).k .attn(1).relu().norm(batch) .loss(huber)` over `recipe.data(\"data/temporal/chronological_splits\").target(\"target\")`. The dataset is ~12000 rows with 1 feature (`input`); arithmetic work is minimal (8-filter kernel-1 conv, 1-head attention, batch norm) and memory traffic is under a few hundred KB, so a defensible CPU epoch should be under 5 s. The 60 s gap reaches the same earliest public failure (`recipe.train().run()` crossing the slow-runtime threshold) through the same CPU native-backend path (`compile_model` -> `NativeTape::epoch`) for a quantized batch-normalized graph covered by issue 622. The `.conv(8, 1)` and `.attn(1)` blocks, `.qi(4).k`, `.norm(batch)`, `.f(6, 9)`, and `.loss(huber)` are configuration differences that execute after the failure point; they expand the existing quantized batch-normalization stall rather than establishing a different earliest cause.",
  "rationale": "The reproduction reaches `recipe.train().run()` > 60 s on `backend=cpu` through the CPU native artifact execution path (`compile_model` / `NativeTape::epoch`), identical to issue 622. The workload (`data/temporal/chronological_splits`, ~12001 rows, 1 feature, width-8 graph, `.qi(4).k` quantization, `.norm(batch)`) justifies sub-5-second throughput; the measured 60 s is materially slower than arithmetic, memory traffic, and CPU capability justify. No different earliest public failure or distinct implementation path is established: `.conv(8, 1)` lowers through the same native emission path as `.layer(8)` in 622, `.attn(1)` adds a small quantized projection, and `.norm(batch)` invokes the same batch-statistics pass. Because the earliest failing operation and inspected native-backend cause match 622, this packet is additional evidence for the existing issue, not a new performance defect."
}
```

## Deterministic failure packet

kind=performance
id=a501e03f4f858e53
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:25176 next:25177 composition:90370963252104
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-1103-25176/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target");
	let model = recipe.model()
		.conv(8, 1).exp().qi(4).k
		.attn(1).relu().norm(batch)
		.loss(huber);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(90370963252104)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Public `recipe.train().run()` stalls past 60 s on `backend=cpu` with `.layer(8).res([conv(8, 1), relu(), layer(8)]).prelu().iq(2).xxs .layer(8).res([conv(8, 1), conv(8, 1)]).sigmoid().norm(batch).iq(2).xxs .loss(mse)` over `recipe.data("data/temporal/chronological_splits").target("target").f(6, 9)` (~9600 training rows, 1 feature `input`). Arithmetic work is minimal: two width-8 residual blocks with kernel-1 convolutions, prelu/sigmoid, and batch normalization with 2-bit integer quantization (`.iq(2).xxs`) yield well under 1 MFLOP per epoch and memory traffic under a few hundred KB. A defensible CPU epoch should be under 5 s, so 60 s is materially slower. Earliest failing operation: `recipe.train().run()` before any epoch completes. Inspected path: CPU native backend (`compile_model` -> `NativeTape::epoch`) for a quantized batch-normalized graph on temporal splits. This reaches the same earliest public failure and same native-backend path as issue 622; the residual blocks, `.prelu()`, `.sigmoid()`, `.iq(2).xxs`, `.loss(mse)`, and `.f(6, 9)` are configuration differences after the failure point and do not establish a different earliest cause.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `d0ce9b66ecd471c5`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 622,
  "title": "Performance: CPU training epoch exceeds 60 s for quantized dense-layer plus kmeans predictor with batch normalization on temporal splits",
  "body": "Public `recipe.train().run()` stalls past 60 s on `backend=cpu` with `.layer(8).res([conv(8, 1), relu(), layer(8)]).prelu().iq(2).xxs .layer(8).res([conv(8, 1), conv(8, 1)]).sigmoid().norm(batch).iq(2).xxs .loss(mse)` over `recipe.data(\"data/temporal/chronological_splits\").target(\"target\").f(6, 9)` (~9600 training rows, 1 feature `input`). Arithmetic work is minimal: two width-8 residual blocks with kernel-1 convolutions, prelu/sigmoid, and batch normalization with 2-bit integer quantization (`.iq(2).xxs`) yield well under 1 MFLOP per epoch and memory traffic under a few hundred KB. A defensible CPU epoch should be under 5 s, so 60 s is materially slower. Earliest failing operation: `recipe.train().run()` before any epoch completes. Inspected path: CPU native backend (`compile_model` -> `NativeTape::epoch`) for a quantized batch-normalized graph on temporal splits. This reaches the same earliest public failure and same native-backend path as issue 622; the residual blocks, `.prelu()`, `.sigmoid()`, `.iq(2).xxs`, `.loss(mse)`, and `.f(6, 9)` are configuration differences after the failure point and do not establish a different earliest cause.",
  "rationale": "The packet is `kind=performance`, `backend=cpu`, `elapsed_seconds:60`, `stable:true`. The dataset (`data/temporal/chronological_splits`, 12001 rows, 1 feature) and model (small quantized residual graph with `.norm(batch)`) justify sub-5-second CPU throughput. The earliest public failure (`recipe.train().run()` exceeding the slow-runtime threshold) is reached through the same CPU native-backend implementation path (`compile_model` / `NativeTape::epoch`) documented in issue 622. No different earliest operation or distinct backend path is established: `.res([...])`, `.conv(8, 1)`, `.prelu()`, `.sigmoid()`, `.iq(2).xxs`, and `.loss(mse)` alter emitted LLVM but do not change the terminal quantized batch-normalization stall. Because issue 622 already tracks CPU stalls for quantized batch-normalized graphs on this dataset, the reproduction expands that issue rather than introducing a new defect."
}
```

## Deterministic failure packet

kind=performance
id=d0ce9b66ecd471c5
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:25301 next:25302 composition:90182863981039
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-1317-25301/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.layer(8)
		.res([conv(8, 1), relu(), layer(8)]).prelu().iq(2).xxs
		.layer(8)
		.res([conv(8, 1), conv(8, 1)]).sigmoid().norm(batch).iq(2).xxs
		.loss(mse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(90182863981039)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Verification receipt: current `minimal` completes every #622 packet under 60 s.

- Base: `minimal` `10148e25` (release build on archy, public CLI `target/release/recipe`, cold = fresh native cache under a new HOME, warm = second run, `ulimit -v 16 GB`, measured with other agents active at the stated load averages)
- `issue-622.rs` on archy cpu: cold 7.69 s, warm 12.63 s; epoch 0.54 s, eval 8.90 s; r2 -4.6130 bce 0.6932 layer.tanh.bnorm.Q2_K/kmeans.silu.bnorm.Q5_0; bundle 304780 B; inference at width 20 = 24.512 (finite), SHA-256 unchanged; load 65.90/78.26
- Inference is a separate `recipe.infer(bundle, &[0.0; WIDTH])` process at the loader-derived width; bundle bytes and SHA-256 are identical before and after.
- No PR is needed for this issue: the packet passes on the current base; closing is your call.

**Comment by nm-z** [comment]
Re-verified on `minimal` b384815c, archy cpu, machine load 24-28, cold native cache per run: `RECIPE_FORCE_CPU=1 recipe reproduction.rs` with `layer(8).tanh().norm(batch).qi(2).k.kmeans(2).silu().norm(batch).qi(5).0` on data/temporal/chronological_splits split 0.8, one epoch: 6.8 s and 7.0 s, loss 0.8005050289784027 both, four native artifacts; a warm 30-epoch run takes 3.1-3.7 s. Margin against the 60 s threshold is 8x; no residual defect on the path, so no new PR.

**Comment by nm-z** [comment]
Closing: no longer reproducible under the measured conditions. Re-verified 2026-09-12 on `minimal` b384815c (archy cpu, load 24-28, cold native cache per run): the packet's reproduction completes one epoch in 6.8-7.0 s (loss 0.8005050289784027 both runs), a warm 30-epoch run in 3.1-3.7 s; 8x margin under the 60 s threshold. b384815c is on current `minimal` 3e278905. Reopen with a packet that crosses 60 s on current `minimal`.
